### ISSUE TITLE
Functions to calculate approximate memory usage needed for compression and decompression

### ISSUE THREAD

**Issue by GregSlazinski** [google/brotli#389]
Is it possible to provide such functions?
Such as :
int64 BrotliCompressionMemUsage(int quality, int lgwin, BrotliEncoderMode mode);
int64 BrotliDecompressionMemUsage(int quality, int lgwin, BrotliEncoderMode mode);

In my program, I use multi-threaded compression, by executing multiple compressors on multiple threads.
However I limit the number of simultaneous threads by the number of available memory .

For example if system has 4 GB of RAM.
And each call to compress function would require 1 GB of memory usage.
Based on that info I would know that I can create up to 4 threads.

However I was unable to find any information about memory usage for compressing and decompressing with Brotli.

Thank you

**Comment by eustas** [comment]
Hello. For decompressor it is easy. For compressor we have only rough estimates (there are about 90 places where memory is allocated).
Going to add method for decompressor soon and for compressor after some code combing...

**Comment by eustas** [comment]
For decompressor the worst case is 2.6MiB + 1.5 * ringbuffer size. Soon an API for calculating ringbuffer size from the first byte will appear + ability to remove 1.5x multiplier.
As for encoder, we still need more time to build the right model.

**Comment by Pro100AlexHell** [comment]
Hello! Could you please answer:
1) Can I use custom allocated buffer for brotli decompression and reuse it, for avoid GC in C#?
2) Which size of memory is need for decompression? May be some table related to compress level?

**Comment by eustas** [comment]
1) Yes, if you write custom memory manager; both encoder and decoder could be initialized with custom memory manager, which could work with preallocated arena
2) In worst case 1.5x LZ window + up to 2MiB for Huffman tables. 1.5x multiplier could be eliminated by appropriate decoder option; normally Huffman tables are much smaller; 2MiB is a theoretical maximum never seen in reality. Normally those are 1-2 orders smaller.
3) Sorry, could you rephrase the question, please

**Comment by Pro100AlexHell** [comment]
3) When the browser receive HTTP header with compression "br", it implement decompression itself. Do you know how good is implementation in firefox and chrome? For example, if my html page need 1000 small files, and all of it is compressed with brotli.. is memory reused? or for each file, the browser create new buffer, 2 MiB (or smaller)? It's browser reated, but I don't even know how to test it.

**Comment by eustas** [comment]
Both Firefox and Chrome use reference (this) decoder.
Decoder does very few allocations, IIRC less than 7. Chrome uses rather effective memory manager (TCMalloc). Combine those facts, and memory reuse would never be a problem.
Can't share exact stats, but in 50+% of cases memory usage is slightly above 32K; in the majority of cases memory usage is below 0.5MiB. That highly depends on the size of the file; to be more precise, it depends on "window size"; Huffman tables together with other all the other overheads is just a tiny fraction of used memory.

**Comment by eustas** [comment]
`BrotliEncoderEstimatePeakMemoryUsage` has been added.
As I've mentioned before - for decoder it is easier, but impossible without knowing the first 2 bytes of input...
